pkg/webhook/webhooks.go

```diff
@@ -0,0 +1,272 @@
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// OpenebsNamespace is default namespace for cstorvolume resources
+	OpenebsNamespace  = "openebs"
+	SourceVolumeLabel = "openebs.io/source-volume"
+)
+
+var (
+	runtimeScheme = runtime.NewScheme()
+	codecs        = serializer.NewCodecFactory(runtimeScheme)
+	deserializer  = codecs.UniversalDeserializer()
+
+	// (https://github.com/kubernetes/kubernetes/issues/57982)
+	defaulter = runtime.ObjectDefaulter(runtimeScheme)
+)
+
+// Skip validation in special namespaces, i.e. in kube-system and kube-public
+// namespaces the validation will be skipped
+var (
+	ignoredNamespaces = []string{
+		metav1.NamespaceSystem,
+		metav1.NamespacePublic,
+	}
+)
+
+// Webhook implements a validating webhook.
+type Webhook struct {
+	//  Server defines parameters for running an golang HTTP server.
+	Server *http.Server
+
+	// kubeClient is a standard kubernetes clientset
+	KubeClient kubernetes.Interface
+
+	// Clientset is a openebs custom resource package generated for custom API group.
+	Clientset clientset.Interface
+}
+
+// WParameters are server configures parameters
+type WParameters struct {
+	// Port is webhook server port
+	Port int
+	//CertFile is path to the x509 certificate for https
+	CertFile string
+	//KeyFile is path to the x509 private key matching `CertFile`
+	KeyFile string
+}
+
+func init() {
+	_ = corev1.AddToScheme(runtimeScheme)
+	_ = admissionregistrationv1beta1.AddToScheme(runtimeScheme)
+	// defaulting with webhooks:
+	// https://github.com/kubernetes/kubernetes/issues/57982
+	_ = v1.AddToScheme(runtimeScheme)
+}
+
+// NewWebhook creates a new instance of a webhook.
+func NewWebhook(p WParameters, kubeClient kubernetes.Interface, openebsClient clientset.Interface) (*Webhook, error) {
+
+	pair, err := tls.LoadX509KeyPair(p.CertFile, p.KeyFile)
+	if err != nil {
+		return nil, err
+	}
+	wh := &Webhook{
+		Server: &http.Server{
+			Addr:      fmt.Sprintf(":%v", p.Port),
+			TLSConfig: &tls.Config{Certificates: []tls.Certificate{pair}},
+		},
+		KubeClient: kubeClient,
+		Clientset:  openebsClient,
+	}
+	return wh, nil
+}
+
+func admissionRequired(ignoredList []string, metadata *metav1.ObjectMeta) bool {
+	// skip special kubernetes system namespaces
+	for _, namespace := range ignoredList {
+		if metadata.Namespace == namespace {
+			glog.Infof("Skip validation for %v for it's in special namespace:%v", metadata.Name, metadata.Namespace)
+			return false
+		}
+	}
+	return true
+}
+
+func validationRequired(ignoredList []string, metadata *metav1.ObjectMeta) bool {
+	required := admissionRequired(ignoredList, metadata)
+	glog.V(4).Infof("Validation policy for %v/%v: required:%v", metadata.Namespace, metadata.Name, required)
```

Comment:
we need to make sure that this log doesn't come by default

Reply:
It will not come util we set the log verbose level 4 or more then that.